### DISCUSSION TITLE
UCLA Miniscope v4 Time Lapse Imaging in Bonsai

### DISCUSSION THREAD

**Discussion by Glia-phile** [bonsai-rx/bonsai#1798] [Q&A]
Hi,

Our group is attempting to use Bonsai to capture images with the UCLA miniscopes v4.  I downloaded the miniscope packages and watched all of the tutorials I could find. I was able to make a simple program to capture time lapse images of a single focal plane. When reviewing the images, we found that the focal plane could shift during longer imaging sessions. We would like to shift from acquiring a single image to a z-stack of images with time lapse imaging. Our goal is for Bonsai to turn on the LED excitation and allow a moment to stabilize, acquire images from deepest to most superficial over a defined range with a defined step size, save the images to an uncompressed video with timestamp, turn off the LED excitation, and then wait 1-2 minutes before repeating the process. We would like the capture process to take 5-10 seconds or less. Does anyone know if Bonsai can do this? I feel like the code should be pretty simple but can't figure it out. Any tips, suggestions, or code are greatly appreciated! I have tried externalizing the EWL property (focal distance) of the miniscope in my current time lapse program, but I am unable to change its value within a loop, and each loop turns the LED excitation on/off which dramatically increases the time to capture the z-stack. Please let me know if you have any questions!

Thanks,
Dan

**Comment by bruno-f-cruz** [comment]
I never used that specific hardware or package, but if you share the workflow i am happy to try to take a look.

**Reply by Glia-phile** [reply]
Thank you!

Screenshot of the marble diagram below...

Miniscope Properties:
   EWL: Sets the focal depth, values -127 to +127... Based on preliminary imaging, we think we want to try to acquire z-stacks from -126 to +126 with a step size of 3 or -125 to +125 with a step size of 5.
   FramesPerSecond: Sets the acquisition rate but also adjusts the shutter speed... lower values produce brighter images. We have found that a value of 20 FPS works well imaging our cells.
   LEDBrightness: Sets the brightness level of the excitation light in the microscope, values 0 (off) to 100 (max)... We have found values between 20 and 50 to work well with out cells
   LEDRespectsTrigger: I don't fully understand this property but think it may be require to accomplish our goal
   SensorGain: Sets the sensitivity of the imaging sensor, values Low, Medium, or High... we have found that a Medium value works well for our cells

The code...
I externalized the UCLA Miniscope v4 EWL property so that I could adjust the focal plane. I tried to feed it a Range, hoping it would proceed through an array of values each iteration but it appears to image the same value each iteration. Currently, the LED turns on and off each iteration... I want the LED excitation light to stay on while capturing a z-stack and turn off between z-stacks. I've tried to use the LEDRespectsTrigger property, but when I have used it the LED does not turn on. The program currently saves images because I haven't figured out how to output video yet. 

Please let me know if you have any questions and thank you again!

![ScreenShot051524](https://github.com/bonsai-rx/bonsai/assets/169836842/f476bb5d-3a32-4159-b941-764b4e03934e)

**Reply by bruno-f-cruz** [reply]
Please add the actual code as asked in #864. Thanks !

**Reply by Glia-phile** [reply]
I'm sorry... Code for the original single plane time lapse and z-stack attached. Thanks again!

[Time Lapse.zip](https://github.com/bonsai-rx/bonsai/files/15327951/Time.Lapse.zip)

**Reply by Glia-phile** [reply]
Quick update, I was able to create a Bonsai node in Visual Studio for time lapse z-stack imaging with miniscopes. The node controls the LED and EWL (focus) of the miniscope and will loop for a user defined number of z-stacks with a user defined delay between z-stack acquisition. The Trigger output signals when Bonsai should capture an image. I am able to generate a series of images but would like to save each individual z-stack as a video file. How do I tell Bonsai to create a new time stamped video file for each z-stack?

My current marble diagram:
![image](https://github.com/user-attachments/assets/0f46bd84-053a-4d28-aece-fa5aae25d9bb)
  
Thanks again,
Dan

**Reply by drish-soneja** [reply]
Hi Dan! I'm also trying to take a z stack in bonsai with miniscope v4. Do you mind sharing your code for TimeWarp? 
Thank you!
